### PR TITLE
feat: Enable more fine-grained digitization configuration

### PR DESCRIPTION
**Pull request by benjaminhuth**
This extends the JSON-format for digitization configuration to allow more fine grained constraints. 

The underlying problem is, that modules in one layer may have a different configuration, but this cannot be reflected by the geometry hierarchy.

## Overview

This changes the JSON plugin in a way, that it allows a list for the `"value"` entry (single entries are still allowed). The configuration gets an additional field `"constraint"` that allows setting some constraints on the modules (for now only restriction in the radial-position of the surface with respect to the coordinate-origin). This can look like this:

```json
{
    "volume": 23,
    "layer": 2,
    "value": [
      {
        "constraint": {
          "r-range": [300.0,320.0]
        },
        "geometric": {  }
      },
      {
        "constraint": {
          "r-range": [500.0,520.0]
        },
        "geometric": {  }
      }
   ]
}
```

In the C++ code this is reflected by an additional object in the `DigiComponentsConfig`, the `DigiConstraint`:

```c++
struct DigiConstraint {
  std::optional<std::pair<double, double>> rRange;

  bool operator()(const Acts::Surface &s,
                  const Acts::GeometryContext &g) const {
    const auto r = std::hypot(s.center(g)[0], s.center(g)[1]);
    if (rRange && (r < rRange->first || r > rRange->second)) {
      return false;
    }

    return true;
  }
};
```

This is then propagated through the `DigitizationAlgorithm` and then checked. 

## Problems

* The `getBoundIndices()` function does not work yet, I'm not sure what it does, but it seems difficult to reflect the constraints here...
* Performance of nested `std::vector`s bad, maybe switch to `boost::small_vector` or something like that?
* Constraints hard-coded

### CODE REVIEW

**Comment by andiwand** [comment]
what I didn't/don't understand is what's the difference between your approach

```
...
{
    "volume": 23,
    "layer": 2,
    "value": [
      {
        "constraint": {
          "r-range": [300.0,320.0]
        },
        "geometric": {...}
      },
      {
        "constraint": {
          "r-range": [500.0,520.0]
        },
        "geometric": {...}
      }
   ]
}
...
```

and something like

```
...
{
    "volume": 23,
    "layer": 2,
    "module": 103,
    "value": {...}
},
{
    "volume": 23,
    "layer": 2,
    "module": 104,
    "value": {...}
}
...
```

File: Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationConfigurator.hpp [hunk 1]
```diff
@@ -62,224 +64,244 @@ struct DigitizationConfigurator {
...
-                outputSegmentation += Acts::BinUtility(
-                    nBins, minPhi, maxPhi, Acts::open, Acts::binPhi);
-              }
-
-            } break;
-
-            default:
-              break;
+      auto dInputConfigVector = inputDigiComponents.find((geoId));
+      if (dInputConfigVector != inputDigiComponents.end()) {
```
Comment (andiwand) [inline Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationConfigurator.hpp:68]: similar as above

```
if (dInputConfigVector == inputDigiComponents.end()) {
  return;
}
```

**Review by andiwand — APPROVED** [review]
:+1: did one walk through but cannot say much about Acts details

File: Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationConfigurator.hpp [hunk 2]
```diff
@@ -62,224 +64,244 @@ struct DigitizationConfigurator {
   void operator()(const Acts::Surface* surface) {
     if (surface->associatedDetectorElement() != nullptr) {
```
[the remark this answers is not in the document]
Comment (benjaminhuth) [reply Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationConfigurator.hpp:65]: Good idea!

**Pull request closed without merging**